plaidml2/edsl/edsl_test.py

```diff
@@ -98,10 +105,338 @@ def arg_max(I):
     O[x0, x2] >= cond(I[x0, x1, x2], Max[x0, x2], IX[x1])
     return as_uint(O, 32)
 
+def sum_over_axis(I):
+    M, N = TensorDims(2)
+    m, n = TensorIndexes(2)
+    I.bind_dims(M, N)
+    O = TensorOutput(N)
+    O[n] += I[m, n]; # contraction
+    return O
+
+def max_over_axis(I):
+    M, N = TensorDims(2)
+    m, n = TensorIndexes(2)
+    I.bind_dims(M, N)
+    O = TensorOutput(N)
+    O[n] >= I[m, n]
+    return O
+
+def matmul(A, B):
+    I, J, K =  TensorDims(3)
+    i, j, k = TensorIndexes(3)
+    A.bind_dims(I, K)
+    B.bind_dims(K, J)
+    C = TensorOutput(I, J)
+    C[i, j] += A[i, k] * B[k, j]
+    return C
+
+def global_min(I):
+    i, j, k = TensorIndexes(3)
+    Neg = -I
+    O_Neg = TensorOutput()
+    O_Neg[()] >= Neg[i, j, k]
+    O = -O_Neg
+    return O
+
+def avg(I):
+    X, Y = TensorDims(2)
+    x, y = TensorIndexes(2)
+    I.bind_dims(X, Y)
+    Sum = TensorOutput()
+    Sum[y] += I[x, y]
+    return Sum / X
+
+def avg_stages(I):
+    X, Y = TensorDims(2)
+    x, y = TensorIndexes(2)
+    I.bind_dims(X, Y)
+    Sum = TensorOutput()
+    Sum[()] += I[x, y]
+    PartialMean = Sum / X
+    return PartialMean / Y
+
+def avg_merge(I):
+    X, Y = TensorDims(2)
+    x, y = TensorIndexes(2)
+    I.bind_dims(X, Y)
+    Sum = TensorOutput()
+    Sum[()] += I[x, y]
+    return Sum / (X * Y)
+
+def skip(I) :
+    M, N = TensorDims(2)
+    i, j = TensorIndexes(2)
+    I.bind_dims(M, N)
+    O = TensorOutput(N)
+    O[2 * i] += I[2 * i, j]
+    return O
+
+def csum(I) :
+    N = TensorDim()
+    i, k = TensorIndexes(2)
+    I.bind_dims(N)
+    O = TensorOutput(N)
+    O[i] += I[k]
+    O.add_constraint(i - k < N)
+    return O
 
 class TestEdsl(unittest.TestCase):
     maxDiff = None
 
+    def test_sum_over_axis(self):
+      I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [1, 784]))
+      O = sum_over_axis(I)
+      program = Program('sum_over_axis', [O])
+     # print(str(program))
```

Comment:
probably remove the various commented out `print(str(program))` lines

Reply:
oops. thanks!